File: apps/athenaeum/packages/web/src/daily-note-id.ts
import * as Schema from "effect/Schema"
import { EntityId } from "@athenaeum/domain"

// "Resolve or create today's note" (task: "a view that resolves/creates 'today's' note ...
// deterministic id from date"). `EntityId` only accepts a ULID or a UUID (node.ts), so — same
// reasoning `tag.ts`'s `BaseTagIds` doc comment already spells out for the Base Tags — a literal
// slug like `"daily-2026-08-20"` isn't a legal id without weakening that schema for every other
// entity, and a real UUIDv5 needs a hashing dependency this app doesn't otherwise carry.
//
// The scheme here reuses the same trick `BaseTagIds` uses (decimal digits are also valid hex
// digits, so a plain calendar date embeds directly into a syntactically-valid UUID): id =
// `00000000-0000-4000-8000-` + the date's `YYYYMMDD` digits, zero-padded to 12 hex characters.
// The `4000-8000` group deliberately differs from `BaseTagIds`' all-zero groups (which are
// `00000000-0000-0000-0000-00000000000N`) purely so the two reserved-id families can never
// collide with each other by construction, not because either group carries real UUIDv4
// version/variant meaning here.
//
// Deterministic per local calendar date: the same device, same day, always resolves to the same
// id — which is exactly what lets the daily-note flow do `getNode(id)` first and only
// `createNode` on `NodeNotFound`, instead of needing a server-side "find-or-create by title"
// query that doesn't exist on this RPC surface.

const pad2 = (n: number): string => String(n).padStart(2, "0")

/** `YYYY-MM-DD` for `date`, in the browser's local timezone (not UTC) — "today" should mean the
 *  user's own calendar day, not whatever the server's clock/timezone happens to be. */
export const localDateStamp = (date: Date): string =>
  `${date.getFullYear()}-${pad2(date.getMonth() + 1)}-${pad2(date.getDate())}`

export const dailyNoteIdForDate = (date: Date): EntityId => {
  const yyyymmdd = `${date.getFullYear()}${pad2(date.getMonth() + 1)}${pad2(date.getDate())}`
  const suffix = yyyymmdd.padStart(12, "0")
  return Schema.decodeUnknownSync(EntityId)(`00000000-0000-4000-8000-${suffix}`)
}

export const dailyNoteTitleForDate = (date: Date): string => `Daily Note — ${localDateStamp(date)}`

export const todayDailyNoteId: EntityId = dailyNoteIdForDate(new Date())
export const todayDailyNoteTitle: string = dailyNoteTitleForDate(new Date())

// --- Retrieval pass (design-review 2026-08-22 finding #1: "no way in the UI to open any note
// except today's") — the day-navigation route (`/notes?date=YYYY-MM-DD`) and the node view both
// need the two inverse operations of the scheme above: "is this stamp a real calendar date?" (to
// validate the query param) and "is this node id a daily note, and for which day?" (so
// `/node/:id` can offer "open in the daily editor"). Both are pure string/date functions on the
// exact same encoding `dailyNoteIdForDate` writes — no new id scheme, no schema change.

const DATE_STAMP_RE = /^(\d{4})-(\d{2})-(\d{2})$/
const DAILY_NOTE_ID_RE = /^00000000-0000-4000-8000-0000(\d{4})(\d{2})(\d{2})$/

/** Parses a `YYYY-MM-DD` stamp into a local-timezone `Date` at midnight — `undefined` for
 *  anything malformed OR non-real (e.g. `2026-02-31`), checked by round-tripping the constructed
 *  `Date` back through its own components (JS `Date` silently rolls invalid days forward). */
export const parseDateStamp = (stamp: string): Date | undefined => {
  const match = DATE_STAMP_RE.exec(stamp)
  if (!match) return undefined
  const year = Number(match[1])
  const month = Number(match[2])
  const day = Number(match[3])
  const date = new Date(year, month - 1, day)
  if (date.getFullYear() !== year || date.getMonth() !== month - 1 || date.getDate() !== day) {
    return undefined
  }
  return date
}

/** The inverse of `dailyNoteIdForDate`: `YYYY-MM-DD` when `id` is a daily-note id whose embedded
 *  digits form a real calendar date, `undefined` for every other id (including ids in the
 *  reserved family with impossible dates — validated via `parseDateStamp` above, not just the
 *  regex, so a hand-crafted `…-000099999999` id never reads as a daily note). */
export const dateStampFromDailyNoteId = (id: string): string | undefined => {
  const match = DAILY_NOTE_ID_RE.exec(id)
  if (!match) return undefined
  const stamp = `${match[1]}-${match[2]}-${match[3]}`
  return parseDateStamp(stamp) === undefined ? undefined : stamp
}

/** `stamp` shifted by `deltaDays` whole days, in local time — drives the prev/next-day controls.
 *  Callers pass a stamp that already went through `parseDateStamp`, so the non-null assertion via
 *  fallback-to-today never fires in practice; it exists to keep this total rather than throwing. */
export const shiftDateStamp = (stamp: string, deltaDays: number): string => {
  const date = parseDateStamp(stamp) ?? new Date()
  date.setDate(date.getDate() + deltaDays)
  return localDateStamp(date)
}

File: apps/athenaeum/native/AthenaeumDomain/Sources/AthenaeumDomain/DailyNoteID.swift
import Foundation

// Swift mirror of `packages/web/src/daily-note-id.ts` — the deterministic-id scheme that lets
// the daily-note flow do `getNode(id)` first and only `createNode` on `NodeNotFound`, instead of
// needing a server-side "find-or-create by title" query that doesn't exist on this RPC surface.
// **Load-bearing for the macOS/iOS app's own task item 1** ("Match the deterministic
// daily-note-id scheme the web client uses... so the SAME workspace/note is addressed from both
// clients — this is what makes cross-client sync testable"): every constant/function below must
// produce byte-identical output to the TS original for the same calendar date, or the two clients
// silently address two different `nodes` rows instead of converging on one.
//
// Placed in `AthenaeumDomain` (not the app-layer `AthenaeumApp` package) even though the TS
// original lives in `packages/web`, not `packages/domain`: this scheme only depends on
// `EntityId`, has zero UI/CF dependencies, and every native surface that wants to address "the
// same daily note" — macOS, iOS, and (per the plan's watchOS quick-capture fallback,
// `native/docs/decisions.md` Decision 2) a future watchOS target that links `AthenaeumDomain` but
// deliberately never links `AthenaeumCore`/Automerge — needs it equally. That's the same
// "zero Cloudflare/React deps, shared by every consumer" test the plan applies to the rest of
// this package's contents.

private func pad2(_ n: Int) -> String {
    n < 10 ? "0\(n)" : String(n)
}

/// `YYYY-MM-DD` for `date`, in the *device's* local calendar/timezone — mirrors `daily-note-id.ts`'s
/// `localDateStamp`: "today" should mean the user's own calendar day, not UTC or the server's
/// clock/timezone.
public func localDateStamp(_ date: Date, calendar: Calendar = .current) -> String {
    let components = calendar.dateComponents([.year, .month, .day], from: date)
    let year = components.year ?? 0
    let month = components.month ?? 0
    let day = components.day ?? 0
    return "\(year)-\(pad2(month))-\(pad2(day))"
}

/// Mirrors `daily-note-id.ts`'s `dailyNoteIdForDate`: `id = "00000000-0000-4000-8000-" +
/// YYYYMMDD` (zero-padded to 12 hex digits — decimal digits are also valid hex digits, the same
/// trick `BaseTagIds` uses). The `4000-8000` group deliberately differs from `BaseTagIds`'
/// all-zero groups purely so the two reserved-id families can never collide by construction.
public func dailyNoteIdForDate(_ date: Date, calendar: Calendar = .current) -> EntityId {
    let components = calendar.dateComponents([.year, .month, .day], from: date)
    let year = components.year ?? 0
    let month = components.month ?? 0
    let day = components.day ?? 0
    let yyyymmdd = "\(year)\(pad2(month))\(pad2(day))"
    let suffix = String(repeating: "0", count: max(0, 12 - yyyymmdd.count)) + yyyymmdd
    // Safe to force-unwrap: the construction above always yields a syntactically valid UUID
    // (fixed-length hex groups), the same guarantee `daily-note-id.ts`'s
    // `Schema.decodeUnknownSync(EntityId)` relies on for its own non-optional return type.
    // swiftlint:disable:next force_try
    return try! EntityId(validating: "00000000-0000-4000-8000-\(suffix)")
}

/// Returns the deterministic daily-note identity for an already validated civil date. This avoids
/// reinterpreting a server-resolved local day through UTC or the device's current time zone.
public func dailyNoteIdForLocalDate(_ localDate: LocalDate) -> EntityId {
    let yyyymmdd = localDate.rawValue.replacingOccurrences(of: "-", with: "")
    let suffix = String(repeating: "0", count: max(0, 12 - yyyymmdd.count)) + yyyymmdd
    // Safe by LocalDate validation plus the fixed UUID prefix.
    // swiftlint:disable:next force_try
    return try! EntityId(validating: "00000000-0000-4000-8000-\(suffix)")
}

/// The inverse of `dailyNoteIdForLocalDate`: returns the encoded civil date only when `id` is a
/// canonical daily-note id whose embedded digits form a real calendar date. This deliberately
/// mirrors the web client's `dateStampFromDailyNoteId` so retrieval surfaces can route a daily
/// note to its typed editor rather than treating it as a legacy page-text preview.
public func localDateFromDailyNoteId(_ id: String) -> LocalDate? {
    let prefix = "00000000-0000-4000-8000-0000"
    guard id.hasPrefix(prefix), id.count == prefix.count + 8 else { return nil }

    let dateDigits = String(id.dropFirst(prefix.count))
    guard dateDigits.allSatisfy(\.isNumber) else { return nil }

    let localDate = "\(dateDigits.prefix(4))-\(dateDigits.dropFirst(4).prefix(2))-\(dateDigits.suffix(2))"
    return try? LocalDate(validating: localDate)
}

/// Mirrors `daily-note-id.ts`'s `dailyNoteTitleForDate`.
public func dailyNoteTitleForDate(_ date: Date, calendar: Calendar = .current) -> String {
    "Daily Note — \(localDateStamp(date, calendar: calendar))"
}

/// Mirrors `daily-note-id.ts`'s `todayDailyNoteId`/`todayDailyNoteTitle`, but as computed
/// properties rather than module-load-time constants: the TS originals are evaluated once when
/// the browser tab's JS module first loads (effectively "app launch"), which is fine for a
/// single-page-load web session but wrong for a long-lived native process that can stay open
/// across a real midnight rollover — recomputing per access is strictly more correct here, not a
/// behavior change for any single app-launch-to-note-resolution flow.
public enum DailyNote {
    public static var todayId: EntityId { dailyNoteIdForDate(Date()) }
    public static var todayTitle: String { dailyNoteTitleForDate(Date()) }
}
